File: 15. 3Sum/15.cpp
/*
sort -> pivot point iteration + two pointer for two sum
time: O(n^2)
space: O(n)
*/

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        set<vector<int>> ans_set;

        sort(nums.begin(), nums.end());
        // assume i < j < k
        for(int i =0; i < nums.size(); i++){
            if (nums[i] > 0){
                break;
            }

            int start = i+1, end = nums.size() - 1;
            while (start < end){
                int cursum = nums[start] + nums[end] + nums[i];
                if (cursum ==0){
                    vector<int> triplet = {nums[start], nums[end], nums[i]};
                    sort(triplet.begin(), triplet.end());
                    ans_set.insert(triplet);
                }
                
                if (cursum < 0 || cursum ==0){
                    while((start+1) < nums.size() && nums[start] == nums[start+1]){
                        start++;
                    } 
                    start++;
                } 
                
                if (cursum > 0 || cursum ==0){
                    while((end - 1) > start && nums[end] == nums[end-1]){
                        end--;
                    }
                    end--;
                }
            }
        }

        vector<vector<int>> ans(ans_set.begin(), ans_set.end());
        return ans;
    }
};
/*
- order of output and order of triplets does not matter
*/

File: 15. 3Sum/15.py
"""
sort + two pointer (two sum)
time: O(n^2)
space: O(1)
"""
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        prevnum = None
        nums.sort()

        for i in range(0, len(nums)-2):
            firstnum = nums[i]

            if firstnum == prevnum:
                continue

            if firstnum > 0:
                break
            
            left = i + 1
            right = len(nums) - 1
            targetsum = 0 - nums[i]

            while left< right:
                sumval = nums[left] + nums[right]

                if sumval == targetsum:
                    # avoid duplicate solutions
                    if nums[left] != nums[left - 1] or (left- 1) == i:
                        res.append([firstnum, nums[left],  nums[right]])
                    right -= 1
                    left += 1
                elif sumval > targetsum:
                    right -= 1
                else:
                    left += 1
            
            prevnum = firstnum

        return res

"""
nums = [-1,0,1,2,-1,-4]
[-4, -1, -1, -1, 2, 2]

res = []
prevnum = None
go through each numi in nums all the way up to third to last number:
    if numi == prevnum:
        continue
    
    use two pointer approach to find two numbers 
    that sum to (0 - numi)
    left = curIndex + 1
    right = end of list
        while left < right:
            
            if sum of left right pointer values > (0 - numi):
                right -= 1
            elif smaller:
                left += 1
            else:
                add solution to res

    prevnum = numi
""" 
